File: Study_250421_CodingTest3/src/com/ssafy/bongki/SWEA_승강제리그.py
class UserSolution:
    def __init__(self):
        self.N = 0
        self.L = 0
        self.leagues = []

    def init(self, N, L, mAbility):
        self.N = N
        self.L = L
        self.M = N // L
        self.leagues = [[] for _ in range(L)]
        for i in range(N):
            ability = mAbility[i]
            league_id = i // self.M
            self.leagues[league_id].append((ability, i))

    def move(self):
        moved = []
        sorted_leagues = [sorted(league, key=lambda x: (-x[0], x[1])) for league in self.leagues]

        ups = [None] * self.L
        downs = [None] * self.L

        for i in range(self.L):
            if i > 0:
                ups[i] = sorted_leagues[i][0]
            if i < self.L - 1:
                downs[i] = sorted_leagues[i][-1]

        for i in range(self.L):
            if ups[i]:
                self.leagues[i].remove(ups[i])
                self.leagues[i - 1].append(ups[i])
                moved.append(ups[i][1])
            if downs[i]:
                self.leagues[i].remove(downs[i])
                self.leagues[i + 1].append(downs[i])
                moved.append(downs[i][1])

        return sum(moved)

    def trade(self):
        moved = []
        sorted_leagues = [sorted(league, key=lambda x: (-x[0], x[1])) for league in self.leagues]

        for i in range(1, self.L):
            top = sorted_leagues[i][0]

            upper_league_sorted = sorted_leagues[i - 1]
            mid_index = (len(upper_league_sorted) + 1) // 2 - 1
            mid = upper_league_sorted[mid_index]

            # 실제 league에서도 remove
            self.leagues[i].remove(top)
            self.leagues[i - 1].remove(mid)

            # swap
            self.leagues[i].append(mid)
            self.leagues[i - 1].append(top)

            moved.extend([top[1], mid[1]])

        return sum(moved)

File: Study_250421_CodingTest3/src/com/ssafy/minji/SWEA_승강제리그.java
package com.ssafy.minji;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

class UserSolution {
	class Team {
		List<Player> players;
		Player maxP;
		Player midP;
		Player minP;
	}

	class Player {
		int number;
		int ability;
	}

	static Team[] league;
	static int n;
	static int L;
	static int mid;

	void init(int N, int L, int mAbility[]) {
		league = new Team[L];
		this.L = L;
		n = N / L;
		mid = n / 2;
		int idx = 0;
		for (int i = 0; i < L; i++) {
			league[i] = new Team();
			league[i].players = new ArrayList<>();

			for (int j = 0; j < n; j++) {
				Player p = new Player();
				p.number = idx;
				p.ability = mAbility[idx++];
				league[i].players.add(p);
			}
			sortAndSetTradable(i);
		}
	}

	int move() {
		int sum = 0;

		for (int i = 0; i <= L - 2; i++) {
			sum += league[i].minP.number;
			sum += league[i + 1].maxP.number;
			league[i].players.remove(league[i].minP);
			league[i].players.add(league[i + 1].maxP);
			league[i + 1].players.remove(league[i + 1].maxP);
			league[i + 1].players.add(league[i].minP);
		}

		for (int i = 0; i < L ; i++) {
			sortAndSetTradable(i);
		}
		
		return sum;
	}

	int trade() {
		int sum = 0;

		for (int i = 0; i <= L - 2; i++) {
			sum += league[i].midP.number;
			sum += league[i + 1].maxP.number;
			league[i].players.remove(league[i].midP);
			league[i].players.add(league[i + 1].maxP);
			league[i + 1].players.remove(league[i + 1].maxP);
			league[i + 1].players.add(league[i].midP);
		}

		for (int i = 0; i < L ; i++) {
			sortAndSetTradable(i);
		}
		
		return sum;
	}

	void sortAndSetTradable(int i) {
		Collections.sort(league[i].players, new Comparator<Player>() {

			@Override
			public int compare(Player o1, Player o2) {
				if(o1.ability == o2.ability) {
					return o2.number - o1.number;
				}
				return o1.ability - o2.ability;
			}
		});

		league[i].maxP = league[i].players.get(n - 1);
		league[i].midP = league[i].players.get(mid);
		league[i].minP = league[i].players.get(0);
	}

}